File: java/src/main/java/com/fabasoad/leetcode/medium/Solution_79.java
package com.fabasoad.leetcode.medium;

public class Solution_79 {

  private static boolean traverse(char[][] board, int x, int y, boolean[][] visited, String word, int ind) {
    if (x >= 0 && x < board.length && y >= 0 && y < board[x].length && !visited[x][y]) {
      final char current = board[x][y];
      if (current == word.charAt(ind)) {
        visited[x][y] = true;
        if (ind + 1 == word.length()) {
          return true;
        }
        return traverse(board, x - 1, y, arrayCopy(visited), word, ind + 1)
            || traverse(board, x + 1, y, arrayCopy(visited), word, ind + 1)
            || traverse(board, x, y - 1, arrayCopy(visited), word, ind + 1)
            || traverse(board, x, y + 1, arrayCopy(visited), word, ind + 1);
      }
    }
    return false;
  }

  private static boolean[][] emptyArray(char[][] board) {
    final boolean[][] visited = new boolean[board.length][];
    for (int j = 0; j < visited.length; j++) {
      visited[j] = new boolean[board[j].length];
    }
    return visited;
  }

  private static boolean[][] arrayCopy(boolean[][] arr) {
    final boolean[][] res = new boolean[arr.length][];
    for (int i = 0; i < res.length; i++) {
      res[i] = new boolean[arr[i].length];
      for (int j = 0; j < arr[i].length; j++) {
        res[i][j] = arr[i][j];
      }
    }
    return res;
  }

  public boolean exist(char[][] board, String word) {
    for (int x = 0; x < board.length; x++) {
      for (int y = 0; y < board[x].length; y++) {
        if (traverse(board, x, y, emptyArray(board), word, 0)) {
          return true;
        }
      }
    }
    return false;
  }

  public static void main(String[] args) {
    final Solution_79 s = new Solution_79();
    assert s.exist(new char[][] {
        {'A','B','C','E'}, {'S','F','C','S'}, {'A','D','E','E'}
    }, "ABCCED");
    assert s.exist(new char[][] {
        {'A','B','C','E'}, {'S','F','C','S'}, {'A','D','E','E'}
    }, "SEE");
    assert !s.exist(new char[][] {
        {'A','B','C','E'}, {'S','F','C','S'}, {'A','D','E','E'}
    }, "ABCB");
    assert s.exist(new char[][] {{'A','B'}, {'C','D'}}, "CDBA");
    assert s.exist(new char[][] {
        {'A','B','C','E'}, {'S','F','E','S'}, {'A','D','E','E'}
    }, "ABCESEEEFS");
  }
}

File: typescript/leetcode/medium/Solution_79.ts
function traverse(board: string[][], word: string, visited: boolean[][], x: number, y: number, i: number): boolean {
  if (word.length === i) {
    return true;
  }
  if (x < 0 || x >= board.length || y < 0 || y >= board[x].length || visited[x][y]) {
    return false;
  }
  visited[x][y] = true;
  if (board[x][y] === word.charAt(i)) {
    if (word.length - 1 == i) {
      return true;
    }
    let a: boolean = traverse(board, word, visited, x + 1, y, i + 1);
    let b: boolean = traverse(board, word, visited, x - 1, y, i + 1);
    let c: boolean = traverse(board, word, visited, x, y + 1, i + 1);
    let d: boolean = traverse(board, word, visited, x, y - 1, i + 1);
    return a || b || c || d;
  }
  return false;
}

function exist(board: string[][], word: string): boolean {
  // Main algorithm
  for (let i = 0; i < board.length; i++) {
    for (let j = 0; j < board[i].length; j++) {
      // Fill visited
      const visited: boolean[][] = []
      for (let i = 0; i < board.length; i++) {
        visited.push([]);
        for (let j = 0; j < board[i].length; j++) {
          visited[i].push(false);
        }
      }
      const res: boolean = traverse(board, word, visited, i, j, 0);
      if (res) return true;
    }
  }
  return false;
}

// console.log('true ->', exist([["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], "ABCCED"))
// console.log('true ->', exist([["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], "SEE"))
// console.log('false ->', exist([["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], "ABCB"))
console.log('true ->', exist([['a','b'],['c','d']], 'cdba'))
